File: frontend/src/modules/users/schema.js
import { z } from "zod";

export const usersSchema = z.object({
    username: z
        .string()
        .min(3, "Username is too short")
        .max(20, "Username is too long")
        .regex(/^[a-zA-Z0-9._]+$/, "Username is not valid"),

    password: z
        .string()
        .min(8, "Password must be at least 8 characters")
        .max(128, "Password is too long")
        .regex(/[A-Za-z\d@$!%*?&#^]{8,128}$/, "Password is not valid")
        .optional(),

    name: z
        .string()
        .min(3, "Name is too short")
        .max(50, "Name is too long")
        .regex(/^[A-Za-z][A-Za-z0-9 ]{3,49}$/, "Name is not valid"),

    email: z
        .string()
        .regex(/^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$/, "Email is not valid")
        .optional(),

    phone: z
        .string()
        .length(10, "Phone Number should be 10 digits")
        .regex(/^[6-9]\d{9}$/, "Invalid Number")
        .optional(),

    role: z
        .string()
        .max(10, "Role is too long")
        .regex(/^(admin|manager|staff|guest)$/, "Invalid Roles")
        .optional(),

    image: z
        .string()
        .max(50, "Image filename is too long")
        .regex(/^[A-Za-z0-9_/-]+\.(jpg|jpeg|png)$/, "Invalid Image")
        .optional(),



});

File: backend/modules/users/schema.py
from datetime import datetime

from pydantic import BaseModel, EmailStr, Field


# Base Schema
class UserBase(BaseModel):
    """
    Shared properties for User schemas.
    """

    username: str = Field(
        ...,
        min_length=3,
        max_length=20,
        pattern=r"^[a-zA-Z0-9._]+$",
    )
    name: str = Field(
        ...,
        min_length=3,
        max_length=50,
        pattern=r"^[A-Za-z][A-Za-z0-9 ]{3,49}$",
    )
    email: EmailStr | None = None
    phone: str | None = Field(
        default=None,
        min_length=10,
        max_length=10,
        pattern=r"^[6-9]\d{9}$",
    )
    role: str = Field(
        default="guest",
        max_length=10,
        pattern=r"^(admin|manager|staff|guest)$",
    )
    image: str | None = Field(
        default=None,
        pattern=r"^[A-Za-z0-9_/\-]+\.(jpg|jpeg|png)$",
    )


# Create Schema
class UserCreate(UserBase):
    """
    Schema for creating a new user.
    """

    password: str = Field(
        ...,
        min_length=8,
        max_length=128,
        pattern=r"[A-Za-z\d@$!%*?&#^]{8,}$",
    )


# Update Schema
class UserUpdate(BaseModel):
    """
    Schema for updating an existing user.
    All fields are optional.
    """

    username: str | None = Field(
        default=None,
        min_length=3,
        max_length=20,
        pattern=r"^[a-zA-Z0-9._]+$",
    )
    name: str | None = Field(
        default=None,
        min_length=1,
        max_length=50,
        pattern=r"^[A-Za-z][A-Za-z0-9 ]{1,49}$",
    )
    email: EmailStr | None = None
    phone: str | None = Field(
        default=None,
        min_length=10,
        max_length=10,
        pattern=r"^[6-9]\d{9}$",
    )
    role: str = Field(
        default="guest",
        max_length=10,
        pattern=r"^(admin|manager|staff|guest)$",
    )
    image: str | None = Field(
        default=None,
        pattern=r"^[A-Za-z0-9_/\-]+\.(jpg|jpeg|png)$",
    )


# Password Update Schema
class UserPasswordUpdate(BaseModel):
    """
    Schema for updating user password
    """

    current_password: str = Field(
        ...,
        min_length=8,
        max_length=128,
        pattern=r"^[A-Za-z\d@$!%*?&#^]{8,}$",
    )
    new_password: str = Field(
        ...,
        min_length=8,
        max_length=128,
        pattern=r"^[A-Za-z\d@$!%*?&#^]{8,}$",
    )


# Read / Response Schema
class UserRead(UserBase):
    """
    Schema returned in API responses.
    """

    id: int
    created_at: datetime
    last_updated: datetime

    class ConfigDict:
        from_attributes = True
